mailpoet/lib/Automation/Integrations/MailPoet/Actions/SendWelcomeEmail/Action.php

```diff
@@ -0,0 +1,134 @@
+    $this->subscribersFinder = $subscribersFinder;
+  }
+
+  public function getKey(): string {
+    return 'mailpoet:send-welcome-email';
+  }
+
+  public function getName(): string {
+    return __('Send Welcome Email', 'mailpoet');
+  }
+
+  public function validate(Workflow $workflow, Step $step, array $subjects = []): ValidationResult {
+    $result = new ValidationResult();
+    if (!isset($step->getArgs()['welcomeEmailId'])) {
+      $result->addError('welcomeEmailIdRequired', 'Step arguments did not include a welcomeEmailId.');
+    } else {
+      $welcomeEmailId = (int)$step->getArgs()['welcomeEmailId'];
+      $newsletter = $this->newslettersRepository->findOneById($welcomeEmailId);
+      if ($newsletter === null) {
+        $result->addError('welcomeEmailNotFound', sprintf("Welcome Email with ID '%s' not found.", $welcomeEmailId));
+      } else {
+        $type = $newsletter->getType();
+        if ($type !== NewsletterEntity::TYPE_WELCOME) {
+          $result->addError('newsletterMustBeWelcomeType', sprintf("Newsletter must be a Welcome Email. Actual type for newsletter ID '%s' was '%s'.", $welcomeEmailId, $type));
+        } else {
+          $result->setNewsletter($newsletter);
+        }
+      }
+    }
+
+    $segmentSubject = $subjects['mailpoet:segment'] ?? null;
+    if (!$segmentSubject instanceof SegmentSubject) {
+      $result->addError('segmentSubjectRequired', "A 'mailpoet:segment' subject is required.");
+    } else {
+      $result->setSegmentSubject($segmentSubject);
+    }
+
+    $subscriberSubject = $subjects['mailpoet:subscriber'] ?? null;
+    if (!$subscriberSubject instanceof SubscriberSubject) {
+      $result->addError('subscriberSubjectRequired', "A 'mailpoet:subscriber' subject is required.");
+    } else {
+      $result->setSubscriberSubject($subscriberSubject);
+    }
+
+    return $result;
+  }
+
+  public function run(Workflow $workflow, WorkflowRun $workflowRun, Step $step): void {
+    $validationResult = $this->validate($workflow, $step, $workflowRun->getSubjects());
+
+    if (!$validationResult->isValid()) {
+      throw InvalidStateException::create()->withErrors($validationResult->getErrors());
+    }
+
+    $newsletter = $validationResult->getNewsletter();
+    $subscriber = $validationResult->getSubscriberSubject()->getSubscriber();
+    if (!$subscriber instanceof SubscriberEntity) {
+      throw InvalidStateException::create();
+    }
+
+    $segment = $validationResult->getSegmentSubject()->getSegment();
+    if (!$segment instanceof SegmentEntity) {
+      throw InvalidStateException::create();
+    }
+
+    // Maybe unnecessary since findSubscribersInSegments is status-aware
+    if ($subscriber->getStatus() !== SubscriberEntity::STATUS_SUBSCRIBED) {
+      throw InvalidStateException::create();
+    }
+
+    $isSubscribedToSegment = $this->subscribersFinder->findSubscribersInSegments([$subscriber->getId()], [$segment->getId()]) !== [];
```

Comment:
This is the check I'd change because the finder does a lot of queries (probably written for some batch jobs). So maybe I'd do something like this:

```php
// inject SubscriberSegmentRepository
$subscriberSegment = $this->subscriberSegmentRepository->findOneBy([
  'subscriber' => $subscriber,
  'segment' => $segment,
  'status' => SubscriberEntity::STATUS_SUBSCRIBED,
]);
```

Sth like this also works but may result in more queries:
```php
$subscriberSegment = $subscriber->getSubscriberSegments()->filter(
  function (SubscriberSegment $item) use ($subscriber) {
    return $item->getSubscriber() === $subscriber && $item->status === SubscriberEntity::STATUS_SUBSCRIBED;
  }
)[0] ?? null;
```

(This way we can also consider omitting the direct filtering by `SubscriberEntity::STATUS_SUBSCRIBED` and check it afterwards to have different message for segment not found vs. wrong status, but that's a tiny issue, I guess.)

Reply:
Thank you so much for this explanation! Using the subscribersFinder definitely seemed like overkill.